Libplanet/Action/AccountStateDeltaImpl.cs

```diff
@@ -1,53 +1,198 @@
+                );
             }
+
+            BigInteger balance = GetBalance(recipient, currency);
+            return UpdateFungibleAssets(
+                _updatedFungibleAssets.SetItem((recipient, currency), balance + amount)
+            );
         }
 
         /// <inheritdoc/>
-        IAccountStateDelta IAccountStateDelta.SetState(Address address, IValue state)
+        [Pure]
+        public IAccountStateDelta TransferAsset(
+            Address sender,
+            Address recipient,
+            Currency currency,
+            BigInteger amount,
+            bool allowNegativeBalance = false
+        )
         {
-            return new AccountStateDeltaImpl(_accountStateGetter)
+            if (amount <= 0)
             {
-                _updatedStates = _updatedStates.SetItem(address, state),
-            };
+                throw new ArgumentOutOfRangeException(
+                    nameof(amount),
+                    "The amount to transfer has to be greater than 0."
+                );
+            }
+
+            BigInteger senderBalance = GetBalance(sender, currency),
+                       recipientBalance = GetBalance(recipient, currency);
+
+            if (!allowNegativeBalance && senderBalance < amount)
+            {
+                var msg = $"The account {sender}'s balance of {currency} is insufficient to " +
+                          $"transfer: {senderBalance} {currency} < {amount} {currency}.";
+                throw new InsufficientBalanceException(sender, currency, 0, msg);
+            }
+
+            return UpdateFungibleAssets(
+                _updatedFungibleAssets
+                    .SetItem((sender, currency), senderBalance - amount)
+                    .SetItem((recipient, currency), recipientBalance + amount)
+            );
+        }
+
+        /// <inheritdoc/>
+        [Pure]
+        public IAccountStateDelta BurnAsset(Address owner, Currency currency, BigInteger amount)
+        {
+            string msg;
+
+            if (amount <= 0)
+            {
+                throw new ArgumentOutOfRangeException(
+                    nameof(amount),
+                    "The amount to burn has to be greater than 0."
+                );
+            }
+            else if (!currency.AllowsToMint(_signer))
+            {
+                msg = $"The account {_signer} has no permission to burn assets of " +
+                      $"the currency {currency}.";
+                throw new CurrencyPermissionException(_signer, currency, msg);
+            }
+
+            BigInteger balance = GetBalance(owner, currency);
+
+            if (balance < amount)
+            {
+                msg = $"The account {owner}'s balance of {currency} is insufficient to burn: " +
+                      $"{balance} {currency} < {amount} {currency}.";
+                throw new InsufficientBalanceException(owner, currency, 0, msg);
```

Comment:
Shouldn't this be the `balance`?
```suggestion
                throw new InsufficientBalanceException(owner, currency, balance, msg);
```